### PR TITLE
Use arbitrary power spectra in LSS sims

### PR DESCRIPTION
**Pull request by jrs65**
This PR adds the ability to use arbitrary power spectra in the LSS sims as well
as adding several precomputed options for a Plankck 2018 cosmology with varying
non-linear corrections.

- refactor(Cosmology): add type annotations to methods
- feat(Cosmology): update default params to Planck 2018
- refactor(cosmology): move growth factor and rate to be methods
- refactor(Cosmology): use same parameter names as CAMB in `from_physical`
- refactor(lsscontainers): move serialisation of Cosmology into a new baseclass
- feat(lsscontainers): containers for power spectra and correlation functions
- feat: update precomputed power spectra
- feat(lss): add ability to configure the input power spectrum

### CODE REVIEW

File: cora/signal/data/generate_powerspectra.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+base_params = {
+    # Initial power spectrum
+    "As": 2.105e-9,
+    "ns": 0.9665,
+    # CAMB params
+    "WantTransfer": True,
+    "WantCls": False,
+    # Matter PS params
+    "redshifts": (1.0,),
+    "kmax": 10.0,
```
Comment (sjforeman) [inline cora/signal/data/generate_powerspectra.py:27]: Why not boost this, just to be safe? `kmax=50` will work without having to change any other camb parameters

File: cora/signal/lss.py [hunk 2]
```diff
@@ -41,41 +40,52 @@ class CalculateCorrelations(task.SingleTask):
         used (see `corrfunc.ps_to_corr` for details). Default: 1
     samples_per_decade : int
         How many samples per decade to calculate. Default: 1000
+    powerspectrum : str
+        If the a power spectrum object is not passed directly to the setup, then this
```
Comment (sjforeman) [inline cora/signal/lss.py:44]: Typo: "the a"

File: cora/signal/lsscontainers.py [hunk 3]
```diff
@@ -229,6 +256,96 @@ def freq(self):
...
+            The redshift to fix the power spectrum at.
+
+        Returns
+        -------
+        psfunc
+            A function which calculates the power spectrum at given wavenumbers.
+        """
+
+        def _ps(k):
+            return self.ps(k, z)
```
Comment (sjforeman) [inline cora/signal/lsscontainers.py:333]: I'm confused when I read this: it looks like `self.ps` should be an array of power spectrum values, but here it is evaluated as a function. Doesn't `self.get_function("powerspectrum")` need to be used here instead?

**Review by sjforeman — APPROVED** [review]
Looks good! Only a few minor questions and comments.

One further comment: should 03cd4509b5eb464fbd0f4c4279167a32666a2b93 be marked as a breaking change? The default cosmological parameters have only shifted by small amounts, but it nevertheless means that some results from previous code versions will be slightly shifted if they are recomputed.

File: cora/signal/lsscontainers.py (hunk 3 above)
Comment (jrs65) [reply cora/signal/lsscontainers.py:333]: Yeah, there might be a bug here. I'm just doing the testing of this all now.

File: cora/signal/data/generate_powerspectra.py (hunk 1 above)
Comment (jrs65) [reply cora/signal/data/generate_powerspectra.py:27]: I could do, the power spectra are all power law extrapolated outside the range of samples, so it just depends if CAMB is going to do anything meaningful in the extended range. I agree it doesn't really hurt though.

**Comment by jrs65** [comment]
I agree it makes sense to mark it with a breaking change.

File: cora/signal/data/generate_powerspectra.py (hunk 1 above)
Comment (jrs65) [reply cora/signal/data/generate_powerspectra.py:27]: I've bumped this all the way to k=1000.

File: cora/signal/lsscontainers.py (hunk 3 above)
Comment (jrs65) [reply cora/signal/lsscontainers.py]: Yup. Bug. Fixed.

**Pull request merged**